src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheckTest.java

```diff
@@ -75,4 +183,38 @@ public void testTokensNotNull() {
             .isNotNull();
     }
 
+    @Test
+    public void testOverloadDescriptorEquals()
```

Comment:
Pure unit test is not much better than suppression, we avoid extension of our technical debt at all costs. Please remove this test to let us see all and work through pitest failures. If code is problematic, remove it and run check regression report in order to find test cases to cover new code.

Reply:
@nick-mancuso I was only able to get 100% coverage by making the equals method
```java
    @Override
     public boolean equals(final Object o) {
         final OverloadDescriptor other = (OverloadDescriptor) o;
         return Objects.equals(pattern.pattern(), other.pattern.pattern())
                && Objects.equals(methodName, other.methodName);
      }
```

However, when I do this I get two PMD errors:

```
[ERROR] Medium: Equals method for com.puppycrawl.tools.checkstyle.checks.coding.OverloadMethodsDeclarationOrderCheck$OverloadDescriptor assumes the argument is of type OverloadMethodsDeclarationOrderCheck$OverloadDescriptor [com.puppycrawl.tools.checkstyle.checks.coding.OverloadMethodsDeclarationOrderCheck$OverloadDescriptor] At OverloadMethodsDeclarationOrderCheck.java:[line 377] BC_EQUALS_METHOD_SHOULD_WORK_FOR_ALL_OBJECTS
[ERROR] High: com.puppycrawl.tools.checkstyle.checks.coding.OverloadMethodsDeclarationOrderCheck$OverloadDescriptor.equals(Object) does not check for null argument [com.puppycrawl.tools.checkstyle.checks.coding.OverloadMethodsDeclarationOrderCheck$OverloadDescriptor] At OverloadMethodsDeclarationOrderCheck.java:[lines 377-378] NP_EQUALS_SHOULD_HANDLE_NULL_ARGUMENT
```

Given the encapsulation of the class, both of the scenarios (comparing for equality vs a null, and comparing for equality against object with a different class) are impossible to occur. Likewise, I also don't have any way of testing those branches without reflection to meet jacoco's requirements of 100% branch coverage. How should I resolve this?

Reply:
We will not extend our technical debt, if we have this much code that we cannot hit, it is time to rethink implementation.

Reply:
This is the only code that doesn't get hit, it is impossible for this scenario to occur. PMD requires this check to be there.
```java
if (o == null || getClass() != o.getClass()) {
                return false;
}
```

I will move the inner class to https://github.com/checkstyle/checkstyle/tree/master/src/main/java/com/puppycrawl/tools/checkstyle/api and make it public so I can add tests for this branch.

Reply:
Actually..I think I can use two maps and get rid of the need for the inner class. Will push up later today